modules/ml/src/main/java/org/apache/ignite/ml/math/primitives/vector/storage/DenseVectorStorage.java

```diff
@@ -56,19 +61,90 @@ public DenseVectorStorage(double[] data) {
         this.data = data;
     }
 
+    /**
+     * @param data Backing data array.
+     */
+    public DenseVectorStorage(Serializable[] data) {
+        assert data != null;
+
+        this.rawData = data;
+    }
+
     /** {@inheritDoc} */
     @Override public int size() {
-        return data == null ? 0 : data.length;
+        if (data == null && rawData == null)
+            return 0;
+        else {
+            if (data != null)
+                return data.length;
+            if (rawData != null)
+                return rawData.length;
+            else
+                throw new IllegalStateException();
+        }
+    }
+
+    /**
+     * Copies values between internal arrays If toNumericRepresentation = true then method tries cast all serializable
+     * objects to double, initializes double array with converted values and cleans raw data array. If
+     * toNumericRepresentation = false then method copies double array to raw data array and cleans double array.
+     *
+     * @param toNumericRepresentation To numeric representation.
+     */
+    private void swap(boolean toNumericRepresentation) {
+        A.ensure(data == null || rawData == null, "data == null || rawData == null");
+        if (data == null && rawData == null)
+            return;
+
+        if (toNumericRepresentation) {
+            if (data != null)
+                return;
+
+            data = new double[rawData.length];
+            for (int i = 0; i < rawData.length; i++)
+                data[i] = ((Number)rawData[i]).doubleValue();
+            rawData = null;
+        } else {
+            if (rawData != null)
+                return;
+
+            rawData = new Serializable[data.length];
+            for (int i = 0; i < rawData.length; i++)
+                rawData[i] = data[i];
+            data = null;
+        }
     }
 
     /** {@inheritDoc} */
     @Override public double get(int i) {
-        return data[i];
+        if (data != null)
+            return data[i];
+
+        Serializable v = rawData[i];
+        if (v == null)
+            return 0.0;
+        else
+            return ((Number)rawData[i]).doubleValue();
+    }
+
+    /** {@inheritDoc} */
+    @Override public <T extends Serializable> T getRaw(int i) {
+        swap(false);
+        return (T)rawData[i];
     }
 
     /** {@inheritDoc} */
     @Override public void set(int i, double v) {
-        data[i] = v;
+        if (data != null)
+            data[i] = v;
+        else
+            rawData[i] = v;
+    }
+
+    /** {@inheritDoc} */
+    @Override public <T extends Serializable> void setRaw(int i, T v) {
+        swap(false);
```

Comment:
Would it be better to keep two keep both arrays "active" and bitset that tells for each index which array contains actual value?

Reply:
No, I don't want to save additional states in vector. From my point of view, vector can be numeric and if we try to save some Serializable object - vector change own state to Serializabled representation. Only "double[] data()" can cast vector to numeric representation. These casts won't be so frequently.